src/arch/arm/32/kernel/vspace.c

```diff
@@ -2214,6 +2214,7 @@ static exception_t decodeARMPageTableInvocation(word_t invLabel, word_t length,
     pd = PDE_PTR(cap_page_directory_cap_get_capPDBasePtr(pdCap));
     asid = cap_page_directory_cap_get_capPDMappedASID(pdCap);
 
+#ifndef CONFIG_ARM_HYPERVISOR_SUPPORT
     if (unlikely(vaddr >= USER_TOP)) {
```

Comment:
USER_TOP prevents mapping entries being created that would overlap with:
1. the kernel window mapping entries, 
2. the kernel device memory mappings
3. the log buffer during bench marking
4. the PD_ASID_SLOT for storing the hardware ASID.

I think item 4. is still needed even in hyp mode and so we still need to prevent mappings from overlapping? 

Another thing to consider is if we need to restrict mapping the very last page of the address space?  I'm not sure what would happen if a syscall was performed on the last address of memory on Arm when the kernel returns after `setNextPC()`

Reply:
PD_ASID_SLOT is still needed. @szhuang Could that error, happened when addressing PT_ASID_SLOT on rk3388, is a result of bypassing this check?

Reply:
That would indeed explain things, well spotted.

It sounds to me like we should make the definition of `USER_TOP` based on `CONFIG_ARM_HYPERVISOR_SUPPORT`, then the check is unchanged, and what exactly it checks can be adjusted. And `USER_TOP` remains a useful concept.

Reply:
Maybe we should give the full 4GiB back to the 32-bit VM? But that means we need to find another place for the ASID_SLOT and verification changes. We just need to make sure that the kernel reserved slots in the stage-2 page directory do not conflict with MMIO addresses that must be used by the VM, so a configurable ASID slot might be enough? Still, I am think about where we could store the ASID.

Reply:
HYP uses long format, and several bits (58-52 and 11-2) in the PDE entry are IGNORED by hardware if the next level is page table. These bits are ignored by hardware, and hardware does not change these bits. They should be enough to contain HW ASID.

Reply:
That sounds possible, actually.

Reply:
Does it also work if there is a page mapped at that level?

Reply:
If the next level is a block, these bits do not exist.  But I do not think 1GiB page size is very useful at the moment.

Reply:
The asid_map_t is an approach that seems to be suitable on aarch64 for solving the same problem. It should also apply here. I made an issue (https://github.com/seL4/seL4/issues/1027) and am going to close this PR for now.